be/src/exec/parquet_builder.cpp

```diff
@@ -0,0 +1,351 @@
+#include <arrow/io/interfaces.h>
+#include <parquet/arrow/writer.h>
+#include <parquet/exception.h>
+
+#include "column/chunk.h"
+#include "column/column_helper.h"
+#include "common/logging.h"
+#include "exprs/column_ref.h"
+#include "exprs/expr.h"
+#include "runtime/exec_env.h"
+#include "util/priority_thread_pool.hpp"
+
+namespace starrocks {
+
+ParquetOutputStream::ParquetOutputStream(std::unique_ptr<WritableFile> writable_file)
+        : _writable_file(std::move(writable_file)) {
+    set_mode(arrow::io::FileMode::WRITE);
+}
+
+ParquetOutputStream::~ParquetOutputStream() {
+    arrow::Status st = ParquetOutputStream::Close();
+    if (!st.ok()) {
+        LOG(WARNING) << "close parquet output stream failed, err msg: " << st.ToString();
+    }
+}
+
+arrow::Status ParquetOutputStream::Write(const std::shared_ptr<arrow::Buffer>& data) {
+    Write(data->data(), data->size());
+    return arrow::Status::OK();
+}
+
+arrow::Status ParquetOutputStream::Write(const void* data, int64_t nbytes) {
+    if (_is_closed) {
+        return arrow::Status::IOError("The output stream is closed but there are still inputs");
+    }
+
+    const char* ch = reinterpret_cast<const char*>(data);
+
+    Slice slice(ch, nbytes);
+    Status st = _writable_file->append(slice);
+    if (!st.ok()) {
+        return arrow::Status::IOError(st.to_string());
+    }
+
+    _cur_pos += nbytes;
+    return arrow::Status::OK();
+}
+
+arrow::Result<int64_t> ParquetOutputStream::Tell() const {
+    return _cur_pos;
+}
+
+arrow::Status ParquetOutputStream::Close() {
+    if (_is_closed) {
+        return arrow::Status::OK();
+    }
+    Status st = _writable_file->close();
+    if (!st.ok()) {
+        LOG(WARNING) << "close parquet output stream failed, err msg: " << st;
+        return arrow::Status::IOError(st.to_string());
+    }
+    _is_closed = true;
+    return arrow::Status::OK();
+}
+
+std::size_t ParquetOutputStream::get_written_len() const {
+    return _cur_pos;
+}
+
+ParquetBuilder::ParquetBuilder(std::unique_ptr<WritableFile> writable_file,
+                               const std::vector<ExprContext*>& output_expr_ctxs, const ParquetBuilderOptions& options)
+        : _writable_file(std::move(writable_file)),
+          _output_expr_ctxs(output_expr_ctxs),
+          _row_group_max_size(options.row_group_max_size) {
+    _init(options);
+}
+
+Status ParquetBuilder::_init(const ParquetBuilderOptions& options) {
+    _init_properties(options);
+    Status st = _init_schema(options.file_column_names);
+    if (!st.ok()) {
+        LOG(WARNING) << "Failed to init parquet schema: " << st;
+    }
+
+    _output_stream = std::make_shared<ParquetOutputStream>(std::move(_writable_file));
+    _buffered_values_estimate.reserve(_schema->field_count());
+    _file_writer = ::parquet::ParquetFileWriter::Open(_output_stream, _schema, _properties);
+    return Status::OK();
+}
+
+void ParquetBuilder::_init_properties(const ParquetBuilderOptions& options) {
```

Comment:
should we set `data_page_version_` to v2?

Reply:
I think we need to set it to optional for user.